BootloaderCommonPkg/Library/TpmLib/TpmLib.c

```diff
@@ -604,33 +682,51 @@ TpmIndicateReadyToBoot (
   @retval RETURN_SUCCESS      Operation completed successfully.
   @retval Others              Unable to extend stage hash.
 **/
+
 RETURN_STATUS
 TpmExtendStageHash (
-  IN       UINT8            ComponentType
+  IN       UINT8            ComponentType,
+  IN       UINT32           Signature
   )
 {
   RETURN_STATUS        Status;
   CONST UINT8         *Digest;
   UINT8                HashAlg;
+  UINT32               Src;
+  UINT32               Length;
+  UINT8                DigestHash[HASH_DIGEST_MAX];
 
   // Get public key hash
   Status = GetComponentHash (ComponentType, &Digest, &HashAlg);
   if(Status == EFI_SUCCESS) {
-    Status = TpmExtendPcrAndLogEvent ( 0, TPM_ALG_SHA256, Digest,
-      EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+    if (HashAlg == PcdGet8(PcdMeasuredBootHashType)){
+
+      Status = TpmExtendPcrAndLogEvent ( 0, (TPMI_ALG_HASH) GetTpmHashAlgEnabled(), Digest,
+        EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+    } else {
+      DEBUG ((DEBUG_INFO, "Calculate Hash for the components\n"));
+      Status = GetComponentInfo (Signature, &Src, &Length);

```

Comment:
Can we consider doing it at the loading time ?   The normal flow will copy it to memory and then do hash calculation, it is more efficient if the component is in flash.

Reply:
Thanks. I would check for doing hash calculation at load time and then extend.

Reply:
Maurice, I will check this optimization as separate patch